### PR TITLE
SymExec: fix data race for concurrent exploration

### PR DESCRIPTION
**Pull request by d-xo**
## Description

Fixes the data race due to the sharing of a `vm` instance to both branches of a call to `concurrently` in the symbolic interpreter. The race was caused by the sharing of mutable memory between threads, fixed by copying memory before passing the vm to each branch.

You can check the fix with the following, which should quickly fail on `main`, but run fine on this branch:

```
while cabal run -f devel test -- -p '/access-out-of-bounds-array/'; do :; done
```

## Checklist

- [x] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [ ] updated the changelog

### CODE REVIEW

**Comment by arcz** [comment]
I think we still need to freeze memory inside frames

**Review by msooseth — APPROVED** [review]
lgtm

**Comment by d-xo** [comment]
> I think we still need to freeze memory inside frames

Ah yes of course, nice catch. Should be fixed now.

File: src/EVM/SymExec.hs [hunk 1]
```diff
@@ -248,6 +250,25 @@ loadSymVM x callvalue cd create =
...
+      mem' <- freeze frame.state.memory
+      pure $ (frame :: Frame Symbolic RealWorld) { state = frame.state { memory = mem' } }
+
+    pure (vm :: VM Symbolic RealWorld)
+      { state = state'
+      , frames = frames'
+      }
+  where
+    freeze = \case
+      ConcreteMemory m -> SymbolicMemory . ConcreteBuf . BS.pack . VUnboxed.toList <$> VUnboxed.freeze m
```
Comment (d-xo) [inline src/EVM/SymExec.hs:269]: This is pretty much a copy paste of `freezeMemory` from `EVM.hs`, but I couldn't figure out how to generalize the type signature there to make it work in both contexts...

**Pull request merged**